### PR TITLE
Add custom metrics locust

### PR DESCRIPTION
**Pull request by kaushikmitr**
added custom metric support for Locust to track metrics like number of tokens sent and received. The framework can be easily extended. 
If the variable `enable_custom_metric` is set to `true` then  custom metrics collected by the locust master is available at the following endpoints:
* While the test is running
```
http://$LOCUST_SERVICE_IP:8089/custom_metrics/custom_metrics.csv
```
* After a test ends:
```
http://$LOCUST_SERVICE_IP:8089/custom_metrics/custom_metrics_final.csv

### CODE REVIEW

File: benchmarks/benchmark/tools/locust-load-inference/locust-docker/locust-tasks/tasks.py [hunk 1]
```diff
@@ -89,6 +99,28 @@ def generate_request(model_params, prompt):
...
+def get_token_count(prompt, resp):
+    """Get number of tokens to prompt and resp using the tokenizer"""
+    global tokenizer
+    backend = model_params["backend"]
+    
+    number_of_input_tokens = len(tokenizer.encode(prompt))
+    number_of_output_tokens = 0
+    
+    if backend == "vllm":
+        number_of_output_tokens = 0 # to be added
```
Comment (puviyarasan) [inline benchmarks/benchmark/tools/locust-load-inference/locust-docker/locust-tasks/tasks.py:111]: What is the difference in output tokens across different model server token backends? Are the outputs not in the resp.content here for other model server backends or is this not in utf-8 format or something else?

File: benchmarks/benchmark/tools/locust-load-inference/locust-docker/locust-tasks/tasks.py [hunk 2]
```diff
@@ -106,15 +140,81 @@ def lm_generate(self):
...
+            tokens_received = -1
+            test_time = -1
+            request_succesful_bool = 0
+            
+            logging.info(f'sending to master: metric_update: {[tokens_sent, tokens_received, test_time, request_succesful_bool]}')
+            self.environment.runner.send_message("metric_update", [tokens_sent, tokens_received, test_time, request_succesful_bool])
+            
+        
+"""
+methods for the locust master to write custome metrics
```
Comment (puviyarasan) [inline benchmarks/benchmark/tools/locust-load-inference/locust-docker/locust-tasks/tasks.py:179]: minor, typo on 'custom'

File: benchmarks/benchmark/tools/locust-load-inference/locust-docker/locust-tasks/tasks.py [hunk 3]
```diff
@@ -106,15 +140,81 @@ def lm_generate(self):
...
-                    f"request {request} failed with: {resp.status_code}")
+            if resp.status_code == 200:
+                self.handle_successful_response(prompt, resp, test_start_time)  
+            else:
+                self.handle_failed_response(request, resp)
+                
+                
+    def handle_successful_response(self, prompt, reponse, start_time):
+        global model_params
+        if model_params['enable_custom_metric'] == 'true':
```
Comment (puviyarasan) [inline benchmarks/benchmark/tools/locust-load-inference/locust-docker/locust-tasks/tasks.py:156]: custom_metric**s**, consider using plural across

File: benchmarks/benchmark/tools/locust-load-inference/locust-docker/locust-tasks/tasks.py (hunk 1 above)
Comment (annapendleton) [reply benchmarks/benchmark/tools/locust-load-inference/locust-docker/locust-tasks/tasks.py:111]: +1

**Review by annapendleton — CHANGES_REQUESTED** [review]

File: benchmarks/benchmark/tools/locust-load-inference/locust-docker/locust-tasks/tasks.py (hunk 1 above)
Comment (achandrasekar) [reply benchmarks/benchmark/tools/locust-load-inference/locust-docker/locust-tasks/tasks.py:111]: The specific API response body is different for different model servers. vLLM, TGI and TensorRT-LLM all are different when it comes to the actual API response. So, we need to handle them individually.

File: benchmarks/benchmark/tools/locust-load-inference/locust-docker/locust-tasks/custom_metric_aggregator.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+    def write_to_csv(self, file_path='custom_metrics.csv'):
+        import csv
+        avg_sent, avg_received, avg_test_time = self.calculate_average_tokens()
+        with open(file_path, mode='w', newline='') as file:
+            writer = csv.writer(file)
+            writer.writerow(['Metric', 'Average Value'])
+            writer.writerow(['# of Successful Req', self.success_count])
+            writer.writerow(['# of Failed Req', self.failure_count])
+            writer.writerow(['Avg Tokens Sent Per Req', avg_sent])
```
Comment (achandrasekar) [inline benchmarks/benchmark/tools/locust-load-inference/locust-docker/locust-tasks/custom_metric_aggregator.py:37]: Thanks for adding these. There are a couple of metrics that are more important and would be good to add:
1. Throughput in output tokens / sec - includes all requests in the run.
2.  Avg request latency in milliseconds per output token

It is ok to add them in a separate change if you'd like.

File: benchmarks/benchmark/tools/locust-load-inference/locust-docker/locust-tasks/tasks.py (hunk 1 above)
Comment (annapendleton) [reply benchmarks/benchmark/tools/locust-load-inference/locust-docker/locust-tasks/tasks.py:111]: I have an implementation for tgi in a WIP PR. I think it's okay for us to implement these other frameworks as a follow up task.

File: benchmarks/benchmark/tools/locust-load-inference/locust-docker/locust-tasks/custom_metric_aggregator.py (hunk 4 above)
Comment (kaushikmitr) [reply benchmarks/benchmark/tools/locust-load-inference/locust-docker/locust-tasks/custom_metric_aggregator.py:37]: yes will do in a follow up PR

**Review by annapendleton — APPROVED** [review]
Thanks @kaushikmitr! Great work!

**Review by achandrasekar — APPROVED** [review]

**Pull request merged**